Fix #1690 -- Bug when exporting tables with symbols on its name

Manerone opened the pull request.

Manerone committed:
Changes to export

- Every table now has an unique identifier before its name
- Every table has their name sanitized

Mytherin commented on src/execution/operator/persistent/physical_export.cpp:
```diff
@@ -46,7 +47,15 @@ static void WriteCopyStatement(FileSystem &fs, stringstream &ss, TableCatalogEnt
-	ss << table->name << " FROM '" << table_file_path << "' (";
+
+	if(KeywordHelper::RequiresQuotes(table->name)){
+		ss << '"'  << table->name << '"' ;
```
Perhaps use `KeywordHelper::WriteOptionallyQuoted`? That would simplify this a bit.

Mytherin commented on test/sql/export/export_database.test:
```diff
@@ -19,9 +19,16 @@ INSERT INTO integers VALUES (1, 3), (4, 2), (NULL, 1)
 statement ok
 CREATE VIEW v1 AS SELECT * FROM integers WHERE i>3; --
 CREATE VIEW v2 AS SELECT * FROM integers WHERE i < 3;
+CREATE VIEW "view.with-symbols" AS SELECT * FROM "table.with-symbols" WHERE i < 3;
```
Could you also test a table/view that has dots and quotes in it (see e.g. test/sql/catalog/test_quoted_column_name.test)?

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some comments:

Also might want to run `make format-fix` to fix the formatting :)

Manerone commented:
I will look into the failing checks tomorrow :)

Mytherin commented:
It looks like the Windows failure is related to the `table_file_path` that is being created for the COPY INTO files (apparently Windows does not like names with a `.` in them). Perhaps you could replace the `.` in the generated file name with `__dot__`?

Manerone commented:
Indeed, from the MS page (https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file) there are a set of characters that are not allowed as file or folder name. I will just replace then by __{NAME}__. I guess this should only be done on the windows side.

Mytherin commented:
Perhaps rather than dealing with that we should name the files `1.csv`, `2.csv`, etc. That way we avoid any potential problems and conflicts arising from that as well.

Mytherin commented on src/include/duckdb/execution/operator/persistent/physical_export.hpp:
```diff
@@ -19,15 +19,17 @@ namespace duckdb {
+	               idx_t estimated_cardinality, vector<CopyInfo> tables_info)
 	    : PhysicalOperator(PhysicalOperatorType::EXPORT, move(types), estimated_cardinality),
-	      function(std::move(function)), info(move(info)) {
+	      function(std::move(function)), info(move(info)), tables_info(tables_info) {
```
move(tables_info)

Mytherin commented on src/include/duckdb/planner/operator/logical_export.hpp:
```diff
@@ -16,11 +16,12 @@ namespace duckdb {
-	LogicalExport(CopyFunction function, unique_ptr<CopyInfo> copy_info)
-	    : LogicalOperator(LogicalOperatorType::LOGICAL_EXPORT), function(function), copy_info(move(copy_info)) {
+	LogicalExport(CopyFunction function, unique_ptr<CopyInfo> copy_info, vector<CopyInfo> tables_info)
+	    : LogicalOperator(LogicalOperatorType::LOGICAL_EXPORT), function(function), copy_info(move(copy_info)), tables_info(tables_info) {
```
move(tables_info)

Mytherin commented on src/planner/binder/statement/bind_export.cpp:
```diff
@@ -13,6 +13,31 @@
 namespace duckdb {
 
+//! Sanitizes a string to have only low case chars and underscores
+string sanitize(const string& str){
```
SanitizeExportIdentifier

Mytherin commented on src/execution/operator/persistent/physical_export.cpp:
```diff
@@ -36,17 +37,12 @@ static void WriteValueAsSQL(stringstream &ss, Value &val) {
-	}
-	ss << table->name << " FROM '" << table_file_path << "' (";
+
+	ss << KeywordHelper::WriteOptionallyQuoted(exported_table.table_name) << " FROM '" << exported_table.file_path
```
I think the table will also need the schema here. Could you add a test case with export in which the table resides in a different schema?

Manerone replied:
It should be taken care already on the bind_export.cpp line 82, this way we avoid having this same `if else` clause at the physical_export.cpp.
```C++
// set up the file name for the COPY TO
if (table->schema->name == DEFAULT_SCHEMA) {
	info->file_path =
			  fs.JoinPath(stmt.info->file_path,
			              StringUtil::Format("%s_%s.%s", to_string(id), SanitizeExportIdentifier(table->name),
			                                 copy_function->function.extension));
} else {
	info->file_path = fs.JoinPath(
			  stmt.info->file_path,
			  StringUtil::Format("%s_%s_%s.%s", SanitizeExportIdentifier(table->schema->name), to_string(id),
			                     SanitizeExportIdentifier(table->name), copy_function->function.extension));
}
```

In any case, I will add the tests because my reasoning might be wrong :)

Manerone replied:
My reasoning was wrong

Mytherin commented on src/planner/binder/statement/bind_export.cpp:
```diff
@@ -8,11 +8,36 @@
+	string result(str);
+
+	for (idx_t i = 0; i < str.length(); ++i) {
+		auto c = str.at(i);
```
`str.at(i)` introduces bounds checks on every access, since we are only accessing until str.length() we know the bounds cannot be exceeded, so you can just use `str[i]`

Mytherin commented on src/planner/binder/statement/bind_export.cpp:
```diff
@@ -8,11 +8,36 @@
+
+		if (c >= 'A' && c <= 'Z') {
+			// To lowercase
+			result.at(i) = tolower(c);
```
Same here, you can use `result[i]`

Mytherin commented on src/execution/operator/persistent/physical_export.cpp:
```diff
@@ -118,8 +114,10 @@ void PhysicalExport::GetChunkInternal(ExecutionContext &context, DataChunk &chun
+	idx_t i = 0;
 	for (auto &table : tables) {
-		WriteCopyStatement(fs, load_ss, (TableCatalogEntry *)table, *info, function);
+		WriteCopyStatement(fs, load_ss, (TableCatalogEntry *)table, *info, exported_tables.at(i), function);
```
The exported_tables seems to rely on the tables scan to output the tables in the same order every time. While that is true currently, that assumption might get violated in the future. What might be better is something along the lines of e.g.

```cpp
struct BoundExportData : public ParseInfo {
    unordered_map<TableCatalogEntry *, ExportedTableData> data;
};
```

Then you have a direct map of table entry -> export data, and you can check that all of the tables are present.

Manerone replied:
Just to be sure I understand, if we gather table info on the bind_export.cpp then we don't need to re-gather the same table info on the physical_export.cpp. Is that correct?

physical_export.cpp:90
```C++
schema->Scan(context.client, CatalogType::TABLE_ENTRY, [&](CatalogEntry *entry) {
	if (entry->type == CatalogType::TABLE_ENTRY) {
		tables.push_back(entry);
	} else {
		views.push_back(entry);
	}
})
```

Mytherin replied:
Yes, indeed. The export lives in a transaction, and a client can only run one query at a time, so necessarily the set of tables visible to a client cannot change in the middle of a query.

Mytherin commented on src/include/duckdb/parser/parsed_data/exported_table_data.hpp:
```diff
@@ -0,0 +1,35 @@
+
+namespace duckdb {
+
+struct ExportedTableData : public ParseInfo {
```
As stated above, I recommend having just a single node (`BoundExportData`) that holds the bind data for all the tables. The BoundExportTable can then inherit from ParseInfo, but the data for each of the individual tables does not need to (i.e. this can just be a regular struct with two string members).

Mytherin reviewed (commented):
Thanks for the updates! Looks great. Some more minor comments, and then I think it's ready to be merged.

Manerone committed:
Adding export tests with different schemas

Manerone committed:
Adding schema info to exported table data

Manerone committed:
Adding schema name to table catalog entry toSQL method

Manerone committed:
Adding export tests with multiple schemas

Manerone committed:
Fixing bugs

- Taking schema name into consideration when exporting
- Using an unordered_map instead of vector to keep exported table info

Manerone committed:
Merge branch 'master' of github.com:duckdb/duckdb

Manerone committed:
tidy check

Manerone committed:
Merge branch 'master' of github.com:duckdb/duckdb

Mytherin commented:
Thanks! Looks great now. One final request, could you add two tables that have a different name but the same lowercase name (e.g. "table" and "TABLE") to the test?

Manerone committed:
Adding more tests

- Adding test to export_database.test where the tables have equal
lowercase names

Mytherin commented:
Thanks! Looks great. Will merge after the tests pass.
